python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,535 @@
+                    list_of_prepared_observations_and_predictions_single_fiber, xd, yd)])
+        except (TypeError, ValueError):
+            pass
+
+        sigle_run_result_for_modification = np.array(sigle_run_result_for_modification)
+        if print_outputs is True:
+            print(sigle_run_result_for_modification)
+        total_penalty_for_single_modification = np.sum(sigle_run_result_for_modification[:, 1])
+        return total_penalty_for_single_modification
+
+    def optimize_function(self, design_variables, return_full_result=False):
+        """Find penalty for all of fibers given the suggested moves.
+
+        This function output needs to be optimized in order to
+        get best fit ot the data.
+
+        Parameters
+        ----------
+        design_variables: `np.array?`
+            variables that describe the transformation
+        return_full_result: `bool`
+            change the amount of output
+
+        Returns
+        ----------
+        if return_full_result == False:
+           single float with total penalty
+        if return_full_result == True:
+           array with all dots, list contaning [index,
+           penalty before optimization, penalty after optimization]
+
+        Notes
+        ----------
+        Takes the default position of black dots and
+        applies simple tranformations to these positions.
+        Calculates the `penalty', i.e., how well do
+        transformed black dots describe the reality.
+
+        Calls `new_position_of_dot`
+        Calls `total_penalty_for_single_dot`
+
+        Examples:
+        ----------
+        # Pass this function to minimization routine.
+        # For November 2021 run I used:
+        >>>> bounds = np.array([(-0.5,0.5), (-0.5,0.5), (-0.005,0.005),
+                           (-0.2,0.2), (-200,200), (-200,200)])
+        # spawn the initial simples to search for the solutions
+        >>>> init_simplex = np.zeros((7, 6))
+        >>>> for l in range(1,7):
+        >>>>     init_simplex[l] = bounds[:, 0]+(bounds[:, 1]-bounds[:, 0])*np.random.random_sample(size=6)
+        >>>> init_simplex[0]=[  0.23196286,  -0.11102326,  -0.00062335,
+                         -0.00675098, -78.35643258,  69.39903929]
+
+        # simple Nelder-Mead will do sufficently good job
+        >>>> res = scipy.optimize.minimize(optimize_black_dots_instance.optimize_function,
+                                      x0=init_simplex[0], method='Nelder-Mead',
+                                      options={'maxiter':1000, 'initial_simplex':init_simplex})
+        """
+        xd_mod = design_variables[0]
+        yd_mod = design_variables[1]
+        scale = design_variables[2]
+        rot = design_variables[3]
+        x_scale_rot = design_variables[4]
+        y_scale_rot = design_variables[5]
+
+        list_of_test_dots = []
+        list_of_total_penalty_for_single_dot = []
+        for i in np.arange(0+1, 2394+1):
```

Comment:
That reads funny. I suggest:

    for i in range(2394):
        fiber = i + 1

(No need to use `np.arange` when you're working with individual values: `range` is more efficient.)